src/main/scala/is/hail/utils/HashMethods.scala

```diff
@@ -0,0 +1,20 @@
+package is.hail.utils
+
+// see e.g. Thorup, "High Speed Hashing for Integers and Strings", for explanation of "multiply-shift" hash functions,
+// both the universal and 2-independent constructions
+class UnivHash32(outBits: Int, factor: Int) extends (Int => Int) {
```

Comment:
Sorry I didn't weigh in sooner on the previous PR.  Two comments:

1.  Randomness.  There should be a global random seed (e.g. 1) and Hail should be deterministic with respect to that random seed.  We should leave the seed as an option, but basically nobody should ever need to change it.  If someone wants to run something many times with different randomized inputs, just running it in succession will achieve the desired effect.  To do this right, we need a random generic per RDD partition determined deterministically from the random seed, but Spark doesn't a convenient place to store this.  I have a partial design for a new abstraction built on RDDs that has per-partition context data that should solve it.  (I also plan to use it for region allocation.)

Cryptographically secure random generator is definitely overkill for us.  I think all randomness should come through `org.apache.commons.math3.random.RandomData`.  I'm OK with a constructor taking `RandomData` to initialize the hash functions.

2. Testing.  Testing is tricky!  I'm not a big fan of unit tests as most people.  You can have a product that passes all unit tests and fails for the user.  I'm a bigger fan of integration tests: that actually tests what the user wants to see.  I haven't looked at the linked paper: do these hash functions have special properties?  Test they satisfy those properties.  If those properties are statistical in nature, compute the result on a lot of values and verify they have the properties you want.  Not necessarily do that in a test because it can be expensive and necessary to verify by hand.  In that case, either document the verification process and/or verify the output a small, fixed set of random inputs so we know if something breaks in the future.  Or, finally, this will be used in a downstream module, and will be tested by the tests for that module.